Deprecated iterator in OctreeLevel.hpp:204

When compiling on a Fedora 37 system using the default clang compiler (details below), I get the following warning:
```
[353/732] Building CXX object axom/axom/quest/CMakeFiles/quest.dir/interface/inout.cpp.o
In file included from /home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:20:
In file included from /home/user/devdir/build/clang-codevelop/axom/include/axom/spin.hpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/DenseOctreeLevel.hpp:13:
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/OctreeLevel.hpp:204:37: warning: 'iterator<std::forward_iterator_tag, const axom::quest::InOutBlockData>' is deprecated [-Wdeprecated-declarations]
  class BlockIterator : public std::iterator<std::forward_iterator_tag, DataType>
                                    ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/detail/inout/InOutOctreeMeshDumper.hpp:375:33: note: in instantiation of template class 'axom::spin::OctreeLevel<2, axom::quest::InOutBlockData>::BlockIterator<const axom::spin::OctreeLevel<2, axom::quest::InOutBlockData>, axom::spin::OctreeLevel<2, axom::quest::InOutBlockData>::ConstBlockIteratorHelper, const axom::quest::InOutBlockData>' requested here
      auto itEnd = levelLeafMap.end();
                                ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/detail/inout/InOutOctreeMeshDumper.hpp:484:38: note: in instantiation of member function 'axom::quest::detail::InOutOctreeMeshDumperBase<2, axom::quest::detail::InOutOctreeMeshDumper<2>>::getBlocksIndexingCell' requested here
    std::vector<BlockIndex> blocks = getBlocksIndexingCell(cIdx);
                                     ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_iterator_base_types.h:127:12: note: 'iterator<std::forward_iterator_tag, const axom::quest::InOutBlockData>' has been explicitly marked deprecated here
    struct _GLIBCXX17_DEPRECATED iterator
           ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/x86_64-redhat-linux/bits/c++config.h:2359:34: note: expanded from macro '_GLIBCXX17_DEPRECATED'
# define _GLIBCXX17_DEPRECATED [[__deprecated__]]
                                 ^
In file included from /home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:20:
In file included from /home/user/devdir/build/clang-codevelop/axom/include/axom/spin.hpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/DenseOctreeLevel.hpp:13:
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/OctreeLevel.hpp:204:37: warning: 'iterator<std::forward_iterator_tag, const axom::quest::InOutBlockData>' is deprecated [-Wdeprecated-declarations]
  class BlockIterator : public std::iterator<std::forward_iterator_tag, DataType>
                                    ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/detail/inout/InOutOctreeMeshDumper.hpp:375:33: note: in instantiation of template class 'axom::spin::OctreeLevel<3, axom::quest::InOutBlockData>::BlockIterator<const axom::spin::OctreeLevel<3, axom::quest::InOutBlockData>, axom::spin::OctreeLevel<3, axom::quest::InOutBlockData>::ConstBlockIteratorHelper, const axom::quest::InOutBlockData>' requested here
      auto itEnd = levelLeafMap.end();
                                ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/detail/inout/InOutOctreeMeshDumper.hpp:703:38: note: in instantiation of member function 'axom::quest::detail::InOutOctreeMeshDumperBase<3, axom::quest::detail::InOutOctreeMeshDumper<3>>::getBlocksIndexingCell' requested here
    std::vector<BlockIndex> blocks = getBlocksIndexingCell(tIdx);
                                     ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_iterator_base_types.h:127:12: note: 'iterator<std::forward_iterator_tag, const axom::quest::InOutBlockData>' has been explicitly marked deprecated here
    struct _GLIBCXX17_DEPRECATED iterator
           ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/x86_64-redhat-linux/bits/c++config.h:2359:34: note: expanded from macro '_GLIBCXX17_DEPRECATED'
# define _GLIBCXX17_DEPRECATED [[__deprecated__]]
                                 ^
In file included from /home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:20:
In file included from /home/user/devdir/build/clang-codevelop/axom/include/axom/spin.hpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/DenseOctreeLevel.hpp:13:
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/OctreeLevel.hpp:204:37: warning: 'iterator<std::forward_iterator_tag, axom::quest::InOutBlockData>' is deprecated [-Wdeprecated-declarations]
  class BlockIterator : public std::iterator<std::forward_iterator_tag, DataType>
                                    ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:639:31: note: in instantiation of template class 'axom::spin::OctreeLevel<2, axom::quest::InOutBlockData>::BlockIterator<axom::spin::OctreeLevel<2, axom::quest::InOutBlockData>, axom::spin::OctreeLevel<2, axom::quest::InOutBlockData>::BlockIteratorHelper, axom::quest::InOutBlockData>' requested here
    auto itEnd = levelLeafMap.end();
                              ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:503:3: note: in instantiation of member function 'axom::quest::InOutOctree<2>::insertMeshCells' requested here
  insertMeshCells();
  ^
/home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:196:18: note: in instantiation of member function 'axom::quest::InOutOctree<2>::generateIndex' requested here
    m_inoutTree->generateIndex();
                 ^
/home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:381:26: note: in instantiation of member function 'axom::quest::internal::InOutHelper<2>::initialize' requested here
    rc = s_inoutHelper2D.initialize(mesh, comm);
                         ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_iterator_base_types.h:127:12: note: 'iterator<std::forward_iterator_tag, axom::quest::InOutBlockData>' has been explicitly marked deprecated here
    struct _GLIBCXX17_DEPRECATED iterator
           ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/x86_64-redhat-linux/bits/c++config.h:2359:34: note: expanded from macro '_GLIBCXX17_DEPRECATED'
# define _GLIBCXX17_DEPRECATED [[__deprecated__]]
                                 ^
In file included from /home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:20:
In file included from /home/user/devdir/build/clang-codevelop/axom/include/axom/spin.hpp:14:
In file included from /home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/DenseOctreeLevel.hpp:13:
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/spin/OctreeLevel.hpp:204:37: warning: 'iterator<std::forward_iterator_tag, axom::quest::InOutBlockData>' is deprecated [-Wdeprecated-declarations]
  class BlockIterator : public std::iterator<std::forward_iterator_tag, DataType>
                                    ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:639:31: note: in instantiation of template class 'axom::spin::OctreeLevel<3, axom::quest::InOutBlockData>::BlockIterator<axom::spin::OctreeLevel<3, axom::quest::InOutBlockData>, axom::spin::OctreeLevel<3, axom::quest::InOutBlockData>::BlockIteratorHelper, axom::quest::InOutBlockData>' requested here
    auto itEnd = levelLeafMap.end();
                              ^
/home/user/devdir/smith/serac/axom/src/axom/core/../../axom/quest/InOutOctree.hpp:503:3: note: in instantiation of member function 'axom::quest::InOutOctree<3>::insertMeshCells' requested here
  insertMeshCells();
  ^
/home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:196:18: note: in instantiation of member function 'axom::quest::InOutOctree<3>::generateIndex' requested here
    m_inoutTree->generateIndex();
                 ^
/home/user/devdir/smith/serac/axom/src/axom/quest/interface/inout.cpp:389:26: note: in instantiation of member function 'axom::quest::internal::InOutHelper<3>::initialize' requested here
    rc = s_inoutHelper3D.initialize(mesh, comm);
                         ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_iterator_base_types.h:127:12: note: 'iterator<std::forward_iterator_tag, axom::quest::InOutBlockData>' has been explicitly marked deprecated here
    struct _GLIBCXX17_DEPRECATED iterator
           ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/x86_64-redhat-linux/bits/c++config.h:2359:34: note: expanded from macro '_GLIBCXX17_DEPRECATED'
# define _GLIBCXX17_DEPRECATED [[__deprecated__]]
                                 ^
4 warnings generated.
```
Here are the details about clang++:
```sh
$ /usr/bin/clang++ --version
clang version 15.0.7 (Fedora 15.0.7-2.fc37)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
This is mainly a heads-up because it's only a warning. It's not holding me back at all.

--- a/src/axom/slam/Map.hpp
+++ b/src/axom/slam/Map.hpp
@@ -575,14 +575,14 @@ public:
     using difference_type = SetPosition;
 
   private:
-    static StackArray<IndexType, Dims + 1> fetchDims(StrideIndexType shape)
+    static StackArray<axom::IndexType, Dims + 1> fetchDims(StrideIndexType shape)
     {
       return {0, shape};
     }
-    static StackArray<IndexType, Dims + 1> fetchDims(
+    static StackArray<axom::IndexType, Dims + 1> fetchDims(
       const StackArray<StrideIndexType, Dims> shape)
     {
-      StackArray<IndexType, Dims + 1> dims;
+      StackArray<axom::IndexType, Dims + 1> dims;
       for(int idim = 0; idim < Dims; idim++)
       {
         dims[idim + 1] = shape[idim];
@@ -595,7 +595,7 @@ public:
       : IterBase(pos)
       , m_map(oMap)
     {
-      StackArray<IndexType, Dims + 1> dataDims = fetchDims(oMap->shape());
+      StackArray<axom::IndexType, Dims + 1> dataDims = fetchDims(oMap->shape());
       dataDims[0] = m_map->size();
       m_mapData =
         axom::ArrayView<DataType, Dims + 1>(m_map->data().data(), dataDims);

--- a/src/axom/slam/SubMap.hpp
+++ b/src/axom/slam/SubMap.hpp
@@ -183,7 +183,7 @@ public:
   ///
 
   /** \brief returns the size of the SubMap  */
-  AXOM_HOST_DEVICE IndexType size() const { return m_subsetIdx.size(); }
+  AXOM_HOST_DEVICE axom::IndexType size() const { return m_subsetIdx.size(); }
 
   /** \brief returns the number of components (aka. stride) of the SubMap  */
   AXOM_HOST_DEVICE IndexType numComp() const

--- a/src/axom/spin/OctreeLevel.hpp
+++ b/src/axom/spin/OctreeLevel.hpp
@@ -201,11 +201,21 @@ public:
    *   \a increment(), \a equal() \a update(), \a pt() and \a data()
    */
   template <typename OctreeLevel, typename IterHelper, typename DataType>
-  class BlockIterator : public std::iterator<std::forward_iterator_tag, DataType>
+  class BlockIterator
   {
   public:
     using GridPt = typename OctreeLevel::GridPt;
     using iter = BlockIterator<OctreeLevel, IterHelper, DataType>;
+
+    // In C++17, inheriting from std::iterator was deprecated.
+    // We provide these typedefs for class BlockIterator to avoid inheriting
+    // from std::iterator and causing warnings for those compiling to C++17
+    // or newer.
+    using value_type = DataType;
+    using difference_type = std::ptrdiff_t;
+    using pointer = DataType*;
+    using reference = DataType&;
+    using iterator_category = std::forward_iterator_tag;
 
     BlockIterator(OctreeLevel* octLevel, bool begin = false)
       : m_octLevel(octLevel)

